paddle/fluid/operators/prim_ops/gather_p_op.cc

```
@@ -0,0 +1,107 @@
+  void RunImpl(const framework::Scope &scope,
+               const platform::Place &dev_place) const override {
+    PADDLE_THROW(platform::errors::Unimplemented(
+        "Prim operator gather_p should not be excuted directly"));
+  }
+};
+
+class GatherPrimOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "(Tensor), The input tensor of gather_p op.");
+    AddInput("IndexTensor",
+             "(Tensor), The index tensor of gather_p op, which is a 1D tensor.")
+        .AsDispensable();
+    AddOutput("Y", "(Tensor), The output tensor of gather_p op.");
+    AddAttr<int64_t>("axis", "(int64_t), The axis along which to gather.");
+    AddAttr<std::vector<int64_t>>(
+        "index", "(std::vector<int64_t>) The index of gather_p op")
+        .SetDefault({0});
+    AddComment(R"DOC(
+Autograd primitive gather_p operator.
+)DOC");
+  }
+};
+
+class GatherPrimOpShapeInference : public framework::InferShapeBase {
+ public:
+  void operator()(framework::InferShapeContext *ctx) const override {
+    framework::InferShapeVarPtr x_var_ptr = ctx->GetInputVarPtrs("X")[0];
+    framework::InferShapeVarPtr y_var_ptr = ctx->GetOutputVarPtrs("Y")[0];
+    int64_t num_index = 0;
+    if (ctx->HasInput("IndexTensor")) {
+      framework::InferShapeVarPtr index_var_ptr =
+          ctx->GetInputVarPtrs("IndexTensor")[0];
+      framework::VarDesc *index_var =
+          BOOST_GET(framework::VarDesc *, index_var_ptr);
+      auto index_shape = index_var->GetShape();
+      PADDLE_ENFORCE_EQ(index_shape.size(), 1,
+                        platform::errors::InvalidArgument(
+                            "The index tensor should be a 1D tensor,"
+                            "but get rank %d",
+                            index_shape.size()));
+      num_index = index_shape[0];
+    } else {
+      num_index = ctx->Attrs().Get<std::vector<int64_t>>("index").size();
+    }
```

Isn't input IndexTensor a mandate? should throw if missing

---

IndexTensor 和 X一样，都是必须存在，框架会检查。